arbitrum/recordingdb.go

```diff
@@ -167,29 +244,94 @@ func PrepareRecording(blockchain *core.BlockChain, lastBlockHeader *types.Header
 		if !lastBlockHeader.Number.IsUint64() {
 			return nil, nil, nil, errors.New("block number not uint64")
 		}
-		recordingChainContext = NewRecordingChainContext(blockchain, lastBlockHeader.Number.Uint64())
+		recordingChainContext = newRecordingChainContext(r.bc, lastBlockHeader.Number.Uint64())
 	}
+	finalDereference = nil
 	return recordingStateDb, recordingChainContext, recordingKeyValue, nil
 }
 
-func PreimagesFromRecording(chainContextIf core.ChainContext, recordingDb *RecordingKV) (map[common.Hash][]byte, error) {
+func (r *RecordingDatabase) PreimagesFromRecording(chainContextIf core.ChainContext, recordingDb *RecordingKV) (map[common.Hash][]byte, error) {
 	entries := recordingDb.GetRecordedEntries()
 	recordingChainContext, ok := chainContextIf.(*RecordingChainContext)
 	if (recordingChainContext == nil) || (!ok) {
 		return nil, errors.New("recordingChainContext invalid")
 	}
-	blockchain, ok := recordingChainContext.bc.(*core.BlockChain)
-	if (blockchain == nil) || (!ok) {
-		return nil, errors.New("blockchain invalid")
-	}
+
 	for i := recordingChainContext.GetMinBlockNumberAccessed(); i <= recordingChainContext.initialBlockNumber; i++ {
-		header := blockchain.GetHeaderByNumber(i)
+		header := r.bc.GetHeaderByNumber(i)
 		hash := header.Hash()
 		bytes, err := rlp.EncodeToBytes(header)
 		if err != nil {
-			panic(fmt.Sprintf("Error RLP encoding header: %v\n", err))
+			return nil, fmt.Errorf("Error RLP encoding header: %v\n", err)
 		}
 		entries[hash] = bytes
 	}
 	return entries, nil
 }
+
+func (r *RecordingDatabase) GetOrRecreateState(ctx context.Context, header *types.Header, logFunc StateBuildingLogFunction) (*state.StateDB, error) {
+	stateDb, err := r.StateFor(header)
+	if err == nil {
+		return stateDb, nil
+	}
+	returnedBlockNumber := header.Number.Uint64()
+	genesis := r.bc.Config().ArbitrumChainParams.GenesisBlockNum
+	currentHeader := header
+	var lastRoot common.Hash
+	for ctx.Err() == nil {
+		if logFunc != nil {
+			logFunc(header, currentHeader, false)
+		}
+		if currentHeader.Number.Uint64() <= genesis {
+			return nil, fmt.Errorf("moved beyond genesis looking for state looking for %d, genesis %d, err %w", returnedBlockNumber, genesis, err)
+		}
+		currentHeader = r.bc.GetHeader(currentHeader.ParentHash, currentHeader.Number.Uint64()-1)
+		if currentHeader == nil {
+			return nil, fmt.Errorf("chain doesn't contain parent of block %d hash %v", currentHeader.Number, currentHeader.Hash())
+		}
+		stateDb, err = r.StateFor(currentHeader)
+		if err == nil {
+			lastRoot = currentHeader.Root
+			defer func() {
+				if (lastRoot != common.Hash{}) {
+					r.dereferenceRoot(lastRoot)
```

Comment:
Let's move this defer out of the for loop to make it a bit more visible and clarify when it'll run